src/docfx/build/xref/XrefMap.cs

```diff
@@ -152,25 +156,71 @@ private bool TryResolveFromExternal(string uid, out XrefSpec spec)
+            // https://github.com/dotnet/corefx/issues/12067
+            // Prefer Dictionary with manual lock to ConcurrentDictionary while only adding
+            Dictionary<string, Lazy<XrefSpec>> map = new Dictionary<string, Lazy<XrefSpec>>();
+            ParallelUtility.ForEach(docset.Config.Xref, url =>
             {
                 var (_, content, _) = RestoreMap.GetRestoredFileContent(docset, url);
                 XrefMapModel xrefMap = new XrefMapModel();
                 if (url?.Value.EndsWith(".yml", StringComparison.OrdinalIgnoreCase) != false)
                 {
                     xrefMap = YamlUtility.Deserialize<XrefMapModel>(content);
+                    foreach (var spec in xrefMap.References)
+                    {
+                        map[spec.Uid] = new Lazy<XrefSpec>(() => spec);
+                    }
                 }
                 else
                 {
-                    xrefMap = JsonUtility.Deserialize<XrefMapModel>(content);
+                    DeserializeAndCreateXrefMap(map, content);
                 }
-                foreach (var spec in xrefMap.References)
+            });
+
+            return new XrefMap(context, map, CreateInternalXrefMap(context, docset.ScanScope));
+        }
+
+        // Assume the input content is 1 line for now since it is what we get from DHS
+        private static void DeserializeAndCreateXrefMap(Dictionary<string, Lazy<XrefSpec>> map, string content)
+        {
+            using (var reader = new StringReader(content))
+            using (var json = new JsonTextReader(reader))
+            {
+                var currentProperty = string.Empty;
+                string uid = null;
+                var startPosition = 1;
+                var endPosition = 1;
+                while (json.Read())
                 {
-                    map[spec.Uid] = spec;
+                    if (json.Value != null)
+                    {
+                        if (json.TokenType == JsonToken.PropertyName)
+                            currentProperty = json.Value.ToString();
+
+                        if (json.TokenType == JsonToken.String && currentProperty == "uid")
+                        {
+                            uid = json.Value.ToString();
+                        }
+                    }
+                    else
+                    {
+                        if (json.TokenType == JsonToken.StartObject)
+                        {
+                            startPosition = json.LinePosition;
+                        }
+                        else if (json.TokenType == JsonToken.EndObject)
+                        {
+                            endPosition = json.LinePosition;
+                            if (uid != null)
+                            {
+                                map[uid] = new Lazy<XrefSpec>(() => JsonUtility.Deserialize<XrefSpec>(content.Substring(startPosition - 1, endPosition - startPosition + 1)));
```

Comment:
will the lazy function take too much spaces since every one has a content string copy?

Reply:
not really, every spec only has their `UID` definition
```json
{"uid":"Microsoft.Office.Interop.Outlook","name":"Microsoft.Office.Interop.Outlook","href":"https://docs.microsoft.com/dotnet/api/microsoft.office.interop.outlook","commentId":"N:Microsoft.Office.Interop.Outlook","fullName":"Microsoft.Office.Interop.Outlook","nameWithType":"Microsoft.Office.Interop.Outlook"}
```